### ISSUE TITLE
Optionally read deployments from MYA certified dir

### ISSUE THREAD

**Issue by slominskir** [JeffersonLab/jmyapi#10]
It would be useful if jmyapi could optionally read deployment config from the MYA certified dir.   At present we currently discover the config is out-of-sync once queries start failing due to changes in host names.

The source of truth for MYA config is found at "/cs/certified/config/mya" on a network drive.   The format of the config there is different (and recently changed) with one file per "deployment" compared to jmyapi (Java API), which uses just a single deployments.properties file for everything and uses Java properties file syntax.

Perhaps we allow the presence of an environment variable such as `MYA_DEPLOYMENT_DIR` to mean ignore deployments.properties and use the deployment config dir instead.  This would require mounting the directory on the host using jmyapi.   A new parser would need to be created that understands how to parse the MYA config dir format.   A file/dir change listener may also need to be added to determine when configs are changed and the ability to reconfigure at runtime automatically.

The major downside is jmyapi is heavily used by myquery, which runs in Tomcat and leverages connection pooling with application server managed DataSources.  That means the above won't work: for each entry in deployments.properties there must be a DataSource configured in Tomcat in the context.xml file.   Documenting this issue here anyways because it comes up every other year and I forget why we didn't do the above.  The whole reconfiguring at runtime piece is also a bit over the top, though a nightly cron restart could be a compromise for 24/7 services such as myquery.

**Comment by slominskir** [comment]
A few incomplete random thoughts:

- One approach may be to create a script run by cron that outputs a new context.xml and deployments.properties and restarts Tomcat once a day.
- Dropping use of Tomcat managed DataSources may be needed (or otherwise may simplify this issue).  Performance impact could be maintained by moving connection pooling to jmyapi as opposed to using Tomcat built-in support.
- An existing host lookup service could be used such as LDAP (Red Hat IdM), or our custom shell command `shownodes 9WS Mya`
- Naming conventions could be followed to map hosts to "deployments"
- DNS and / or DNS CNAMEs?
- Container orchestration software handles this kind of stuff already and container names are often randomly created, but a router service routes traffic to correct host.

**Comment by slominskir** [comment]
The assumption early on was that MYA databases were fairly persistent.   Now it appears they are frequently changing and apparently one at a time in a rolling fashion so this issue ticket is now more relevant.

Recap of this issue:

- Mounting the network file system path containing mya config isn't an issue
- Since we are running an always on server, we would need to know when to take downtime and re-configure and restart.  Not a huge deal to just do a nightly cron restart of the web server for re-configure, but not awesome either.  Monitoring for file timestamp changes could make things better.
- A bigger obstacle is that the Tomcat web server provides database caching via a Tomcat specific config file.   We would need to write a script that parses the mya config and then outputs a Tomcat config.  Alternatively we could drop database connection pooling or move it out of Tomcat and into the app itself.
- The app itself also needs to know the config separately from the Tomcat database connection pool config.  This means another parser would need to be written.  Or just this one parser if we remove connection pooling from Tomcat.

**Comment by apcarp** [comment]
Tomcat and the app seem to transparently pickup the changes.  The last two times I've updated the deployment.properties and context.xml configs, I did not need to do anything besides change the file contents.  I can test this more thoroughly if you have reasons to believe that more is needed.

If the above works fine, then I would assume have a cronjob on our config server that compares the cached configs against a parsed version of /cs/certified/config/mya/*.  If a change is made, copy out the updated config files.
